File: dbeam-core/src/main/scala/com/spotify/dbeam/MetricsHelper.scala
/*
 * Copyright 2017 Spotify AB.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.spotify.dbeam

import org.apache.beam.sdk.PipelineResult
import org.apache.beam.sdk.metrics._

import scala.collection.JavaConverters._
import scala.util.Try

@SuppressWarnings(Array("org.wartremover.warts.AsInstanceOf",
  "org.wartremover.warts.TraversableOps"))
object MetricsHelper {


  def getMetrics(result: PipelineResult): Map[String, Long] = {
    val metrics: MetricQueryResults = result.metrics().queryMetrics(MetricsFilter.builder().build())

    val gauges = metricsAtSteps(metrics.getGauges.asScala)
        .map{ case (k: MetricName, v: Map[String, MetricValue[GaugeResult]]) => (k.getName(),
        v.values.map(_.committed.getOrElse(GaugeResult.empty())).reduce(
          (x: GaugeResult, y: GaugeResult) =>
            if (x.getTimestamp isAfter y.getTimestamp) x else y).getValue
        )}
    val counters = metricsAtSteps(
      metrics.getCounters.asScala.asInstanceOf[Iterable[MetricResult[Long]]])
      .map{
        case (k: MetricName, v: Map[String, MetricValue[Long]]) =>
          (k.getName(), reduceMetricValues(v))}
    (gauges.toSeq ++ counters.toSeq).toMap
  }

  private def metricsAtSteps[T](results: Iterable[MetricResult[T]])
  : Map[MetricName, Map[String, MetricValue[T]]] =
    results
      .groupBy(_.getName())
      .mapValues { xs =>
        val m: Map[String, MetricValue[T]] = xs.map { r =>
          r.getStep -> MetricValue(r.getAttempted, Try(r.getCommitted).toOption)
        } (scala.collection.breakOut)
        m
      }

  private def reduceMetricValues(xs: Map[String, MetricValue[Long]]) = {
    xs.values.foldLeft(0L)( (t, m) => t + m.committed.getOrElse(0L))
  }

  final case class MetricValue[T](attempted: T, committed: Option[T])
}

File: dbeam-core/src/main/java/com/spotify/dbeam/beam/MetricsHelper.java
/*-
 * -\-\-
 * DBeam Core
 * --
 * Copyright (C) 2016 - 2018 Spotify AB
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * -/-/-
 */

package com.spotify.dbeam.beam;

import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;
import java.util.function.BinaryOperator;
import java.util.function.Function;
import java.util.function.ToLongFunction;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;
import org.apache.beam.sdk.PipelineResult;
import org.apache.beam.sdk.metrics.GaugeResult;
import org.apache.beam.sdk.metrics.MetricQueryResults;
import org.apache.beam.sdk.metrics.MetricResult;
import org.apache.beam.sdk.metrics.MetricsFilter;

public class MetricsHelper {

  private static final Function<MetricResult<GaugeResult>, GaugeResult> GET_COMMITTED_GAUGE =
      metricResult -> {
        try {
          return metricResult.getCommitted();
        } catch (UnsupportedOperationException e) {
          return GaugeResult.empty();
        }
      };

  public static final ToLongFunction<MetricResult<Long>> GET_COMMITTED_COUNTER =
      metricResult -> {
        try {
          return metricResult.getCommitted();
        } catch (UnsupportedOperationException e) {
          return 0L;
        }
      };

  public static Map<String, Long> getMetrics(final PipelineResult result) {
    final MetricQueryResults metricQueryResults =
        result.metrics().queryMetrics(MetricsFilter.builder().build());

    final Map<String, Long> gauges =
        StreamSupport.stream(metricQueryResults.getGauges().spliterator(), false)
            .collect(
                Collectors.groupingBy(
                    MetricResult::getName,
                    Collectors.reducing(
                        GaugeResult.empty(),
                        GET_COMMITTED_GAUGE,
                        BinaryOperator.maxBy(Comparator.comparing(GaugeResult::getTimestamp)))))
            .entrySet()
            .stream()
            .collect(Collectors.toMap(e -> e.getKey().getName(), e -> e.getValue().getValue()));

    final Map<String, Long> counters =
        StreamSupport.stream(metricQueryResults.getCounters().spliterator(), false)
            .collect(
                Collectors.groupingBy(
                    m -> m.getName().getName(), Collectors.summingLong(GET_COMMITTED_COUNTER)));
    Map<String, Long> ret = new HashMap<>();
    ret.putAll(gauges);
    ret.putAll(counters);
    addCalculatedMetrics(counters, ret);
    return Collections.unmodifiableMap(ret);
  }

  private static void addCalculatedMetrics(final Map<String, Long> counters,
                                           final Map<String, Long> ret) {
    // calculate and add KBps
    final Long writeElapsedMs = counters.get("writeElapsedMs");
    final Long kbps;
    if (writeElapsedMs != null && writeElapsedMs > 0) {
      kbps = counters.get("bytesWritten") / writeElapsedMs;
    } else {
      kbps = -1L;
    }
    ret.put("KbWritePerSec", kbps);
  }
}
